builtin/providers/aws/resource_aws_db_instance.go

```diff
@@ -605,11 +611,11 @@ func resourceAwsDbInstanceDelete(d *schema.ResourceData, meta interface{}) error
 
 	opts := rds.DeleteDBInstanceInput{DBInstanceIdentifier: aws.String(d.Id())}
 
-	finalSnapshot := d.Get("final_snapshot_identifier").(string)
-	if finalSnapshot == "" {
-		opts.SkipFinalSnapshot = aws.Bool(true)
-	} else {
-		opts.FinalDBSnapshotIdentifier = aws.String(finalSnapshot)
+	skipFinalSnapshot := d.Get("skip_final_snapshot").(bool)
+	opts.SkipFinalSnapshot = aws.Bool(skipFinalSnapshot)
+
+	if name, present := d.GetOk("final_snapshot_identifier"); present && !skipFinalSnapshot {
+		opts.FinalDBSnapshotIdentifier = aws.String(name.(string))
 	}
```

Comment:
If the user has `skip_final_snapshot = false` in their config, but omits `final_snapshot_identifier` (valid config as we have things now), then no final snapshot will be taken. We should error in this case.